compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -97,6 +98,31 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         Ok(())
     }
 
+    fn expose_if_needed(
+        &mut self,
+        src: &ImmTy<'tcx, M::PointerTag>,
+        cast_ty: Ty<'tcx>,
+    ) -> InterpResult<'tcx> {
+        // If we are casting from a pointer to something
+        // that is not a pointer, expose the address of the pointer
+        if src.layout.ty.is_any_ptr() && !cast_ty.is_any_ptr() {
+            let scalar = src.to_scalar()?;
+            let ptr = self.scalar_to_ptr(scalar);
```

Comment:
We do all the `to_scalar` stuff also in `misc_cast`, so wouldn't it make more sense to also do this part there?

Somewhere it should also detect the other direction and give the machine a chance to treat that specially, too. After all we want to attach a wildcard provenance to pointers cast from integers, but *not* to pointers *transmuted* from integers.

I'm thinking we could just have machine hooks `cast_ptr_to_int` and `cast_int_to_ptr` that fully control what happens here (i.e., they can also do arbitrary transformations of the value).

Reply:
This is here because I wanted to make `expose_ptr` take `&mut self`. Is that something that's a good idea, or should it just take `&self`?

Reply:
It should probably take `&mut self`. You can change `misc_cast` to take `&mut self`.

Reply:
Oh right, I tried that, but `misc_cast` is also used here in `read_discriminant`. I don't know what semantics those reads have, but I don't think it should expose pointers.

https://github.com/rust-lang/rust/blob/fbdb10f9fabe47eb763cb4b52b5721740cc63783/compiler/rustc_const_eval/src/interpret/operand.rs#L708

Reply:
Ah, good point. That's annoying...

So maybe we should make `cast_from_int_like` `pub(super)`, and change its signature slightly so that it takes a `Scalar` rather than a `u128`. Then `read_discriminant` can call `cast_from_int_like` and we can make `misc_cast` take `&mut self`.